### DISCUSSION TITLE
How to specify to display a fixed file when the user opens theia for the first time

### DISCUSSION THREAD

**Discussion by LionLion001** [eclipse-theia/theia#11938] [Q&A]

**Comment by msujew** [comment]
Hey @XiaoXuan001,

the `FrontendApplication` class contains a `createDefaultLayout` method which you can override to determine the widget layout when the user opens the app for the first time. You can use it like this:

```ts
@injectable()
export class CustomFrontendApplication extends FrontendApplication {

    @inject(WorkspaceService) private readonly workspaceService: WorkspaceService;
    @inject(FileService) private readonly fileService: FileService;
    @inject(EditorManager) private readonly editorManager: EditorManager;

    protected override async createDefaultLayout(): Promise<void> {
        // Initialize the default layout for search-in-workspace, git, etc.
        await super.createDefaultLayout();

        // Open app.py if it exists
        const appPy = this.workspaceService.getWorkspaceRootUri(undefined)?.resolve('app.py');
        if (appPy && await this.fileService.exists(appPy)) {
            const editor = await this.editorManager.getOrCreateByUri(appPy, { preview: false });
            await this.shell.addWidget(editor, { area: 'main' });
        }
    }
}
```

**Reply by LionLion001** [reply]
thanks for your guidance! If I want to create an extension, how can I pass the file path parameter at startup to tell the program which file to open

**Reply by msujew** [reply]
@XiaoXuan001 You can create a new backend `CliContribution` (see [here](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/node/cli.ts)) and store the argument in there. Then, you proxy the created backend service to the frontend (see [here](https://theia-ide.org/docs/json_rpc/)) and call a function which retrieves the stored argument.

**Reply by LionLion001** [reply]
Can this file path parameter be passed through the url accessed by the browser?

**Reply by msujew** [reply]
Sure, you can just use the snippet I've posted initially and replace `'app.py'` with the [`URL.searchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URL/searchParams) that you like.

**Reply by LionLion001** [reply]
Thanks for your answer. I have used your method to implement it, but it seems that this method can only take effect at the first visit. If I want users to take effect every time they visit the browser, which method should I use

**Reply by msujew** [reply]
If you want this to take effect every time the user opens the IDE, you should create your own [`FrontendApplicationContribution`](https://github.com/eclipse-theia/theia/blob/ac7a9ff51afa71d26673a7a8f7968c8e9a3f51ff/packages/core/src/browser/frontend-application.ts#L34-L35) and simply put the same code as above in the `onStart` method. That will run every time the frontend starts.

Note that your initial question just concerned itself with the event that the user starts the IDE for the **first time**.

**Reply by LionLion001** [reply]
But this code doesn't seem to work in the onStart method .I'm not familiar with these functions
` const appPy = this.workspaceService.getWorkspaceRootUri(undefined)?.resolve('app.py');
        if (appPy && await this.fileService.exists(appPy)) {
            const editor = await this.editorManager.getOrCreateByUri(appPy, { preview: false });
            await this.shell.addWidget(editor, { area: 'main' });
        }`

**Reply by msujew** [reply]
These should work the same way as they do in the `CustomFrontendApplication`, you just need to inject the correct services for them.

**Reply by schattmare** [reply]
Hi! I wanted to try this too, and I created a file "display-file-contibution.ts" where I put in the above code snippet. Then I tried to make the example run following the pattern of the "Hello-world-contibution" (sorry I am a total newbie). I figured I need to create some kind of binding, so I created a file "display-file-frontend-module.ts" with the content "import { DisplayFileFrontendApplication } from './display-file-contribution';" This didn´t work; I suppose for an "extends" a different binding scheme is needed as opposed to a CommandContribution that 'implements'.  Could you provide the code snippet for binding this CustomFrontendApplication? Any help is very  appreciated, thanks in advance.

**Reply by vince-fugnitto** [reply]
@schattmare are you still experiencing issues?
